proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+
+Local modules
+~~~~~~~~~~~~~
+
+We propose adding a language extension ``LocalModules``. This allows
+module definitions within modules, with the same syntax as in [ghc283]_. Local
+modules are very like file modules, in particular:
+
+- They can have imports.
+
+  - The imports can only target modules which can be referred to
+    with a qualified name in the current scope. Their target module name is *never*
+    interpreted as a dotted name.
+
+- They can have nested module definitions.
+- They can have export specifiers.
+- They can *not* have dotted names.
+
+The rules for local modules are mostly the same as for file modules,
+but:
+
+- A local module definition adds a binding for its declared name to
+  its enclosing scope, as usual for definitions.
+- Bindings from the
+  enclosing scope are in scope in the local module, but are shadowed if
+  there is a local definition of the same name. [#]_
+
+.. [#] See the `Recursion`_ section for some subtleties.
+
+Local modules *cannot* import file modules. This ensures that we can
+always compute the dependency graph of the compilation units of a
+Haskell program by looking at the top-level imports alone. [#]_
+
+.. [#] See `Do we need different syntax for "local" imports?`_ for discussions
+   of the consequences.
+
+Module aliases
+~~~~~~~~~~~~~~
+
+We propose adding a language extension ``ModuleAliases``. This allows a
+new kind of top-level definition: ``module <name> = <qualname>``
+
+A module alias adds a binding in the enclosing scope from the new name
+to the module referred to by the right hand side.
+
+Examples
+--------
+
+Exporting small utility modules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+::
+
+   module A (f, module Unsafe) where
+
+   f = Unsafe.g …
+
+   module Unsafe where
+       g = … 
+
+Exporting qualified names
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal allows exporting modules, but doesn’t say they can be
+exported under *qualified* names, which is something that people want to
+do. Consider the following `proposed
+example <https://github.com/ghc-proposals/ghc-proposals/pull/283#issuecomment-541525245>`_:
+
+::
+
+   module MyPrelude ( qualified module BL
+                    , qualified module BS
+                    , Set
+                    , qualified module Set ) where
+
+
+   import qualified Data.ByteString.Lazy as BL
+   import qualified Data.ByteString as BS
+   import Data.Set ( Set )
+   import qualified Data.Set as Set
+
+This could be achieved as follows:
+
+::
+
+   module MyPrelude (module Data, Set) where
+
+   import qualified Data.ByteString.Lazy as BL
+   import qualified Data.ByteString as BS
+   import Data.Set ( Set )
+   import qualified Data.Set as Set
+
+
+   module Data (module ByteString, module Set) where
+       module ByteString (module Lazy, module BS (..)) where
+           import BS
+           module Lazy (module BSL (..)) where
+               import BSL
+       module Set (module Set (..)) where
+           import Set
```

Comment:
I'm not sure that the examples correspond to each other:

The former example doesn't export any module prefixed with `Data`, but does export modules named `BS` and `BL`.

Reply:
It looks like I just misunderstood what the first example was trying to achieve. Let me fix that.

Reply:
I think I've got it right now.